package/src/core/define-integration.ts

```diff
@@ -0,0 +1,89 @@
+import type { AstroIntegration, HookParameters } from "astro";
+import { defu } from "defu";
+import type { AnyOptions, AnyPlugin, ExtendedHooks } from "./types.js";
+import { DEFAULT_HOOK_NAMES } from "../internal/constants.js";
+
+/**
+ * A powerful wrapper around the standard Astro Integrations API. It allows to provide extra hooks, functionality
+ * and best-practices when creating Astro Integrations.
+ *
+ * @param {object} params
+ * @param {string} params.name - The name of your integration
+ * @param {object} params.object - Any default config options you want to set
+ * @param {function} params.setup - This will be called from your `astro:config:setup` call with the user options
+ *
+ * @see https://astro-integration-kit.netlify.app/utilities/define-integration/
+ *
+ * @example
+ * ```ts
+ * export default defineIntergration({
+ *		name: "my-integration",
+ *		defaults: {
+ *    		foo: "bar",
+ * 		},
+ * 		setup({ options }) {
+ * 			console.log(options.foo); // "bar"
+ * 		}
+ * })
+ * ```
+ */
+export const defineIntegration = <
+	TOptions extends AnyOptions = never,
+	TPlugins extends Array<AnyPlugin> = [],
+>({
+	name,
+	options: optionsDef,
+	setup,
+	plugins: _plugins,
+}: {
+	name: string;
+	options?: TOptions;
+	plugins?: TPlugins;
+	setup: (params: {
+		name: string;
+		options: TOptions;
+	}) => ExtendedHooks<TPlugins>;
+}): ((options?: TOptions["options"]) => AstroIntegration) => {
+	return (_options?: TOptions["options"]) => {
+		const options = defu(
+			_options ?? {},
+			optionsDef?.defaults ?? {},
+		) as TOptions["options"];
+
+		const resolvedPlugins = Object.values(
+			(() => {
+				const plugins: Record<string, AnyPlugin> = {};
+				for (const plugin of _plugins ?? []) {
+					plugins[plugin.name] = plugin;
+				}
+				return plugins;
+			})(),
+		);
+
+		const providedHooks = setup({ name, options });
+
+		const hooks: AstroIntegration["hooks"] = Object.fromEntries(
+			DEFAULT_HOOK_NAMES.map((hookName) => [
+				hookName,
+				(params: HookParameters<typeof hookName>) => {
+					const plugins = resolvedPlugins.filter((p) => p.hook === hookName);
```

Comment:
Is there no way for a plugin to interfere with more than one hook?
This would enable things like post-processing on `astro:build:done` based on values from `astro:config:setup` or `astro:config:done` like it is done for Pagefind in Starlight

Reply:
I know we talked about this but consider this PR as a first step! I'm open to adding support for this in a later PR, I had some ideas about how to do that but just decided to postpone it. I'll create some issues to track the progress on upcoming features